src/Compilers/CSharp/Portable/Lowering/AsyncRewriter/AsyncMethodBuilderMemberCollection.cs

```diff
@@ -250,6 +264,42 @@ internal static bool TryCreate(SyntheticBoundNodeFactory F, MethodSymbol method,
             throw ExceptionUtilities.UnexpectedValue(method);
         }
 
+        private static NamedTypeSymbol ValidateBuilderType(SyntheticBoundNodeFactory F, object builderAttributeArgument, Accessibility desiredAccessibility, int desiredArity)
+        {
+            var builderType = builderAttributeArgument as NamedTypeSymbol;
+
+            bool isArityOk;
+            if (desiredArity == 0)
+            {
+                isArityOk = builderType?.IsGenericType == false;
+            }
+            else if (desiredArity == 1)
+            {
+                isArityOk = builderType?.IsUnboundGenericType == true &&
+                            builderType?.ContainingType?.IsGenericType != true;
+            }
+            else
+            {
+                isArityOk = false;
+                Debug.Assert(false, "expected arity 0 or 1");
+            }
+
+
+            if ((object)builderType == null ||
+                 builderType.IsErrorType() == true ||
+                 builderType.SpecialType == SpecialType.System_Void ||
+                 builderType.DeclaredAccessibility != desiredAccessibility ||
```

Comment:
Checking the accessibility matches the task-like type accessibility exactly seems a little surprising. Ideally we'd want to check accessibility of `builderType` with respect to the caller, so we could allow an `internal` builder type on a `public` task-like type for instance.